Notification of significant events during bag recording and playback

gbiggs opened the pull request:
Closes #884

gbiggs commented:
I could do with some suggestions on how best to test this functionality.

jhdcs commented:
At the moment, it's looking like `rosbag2_transport` doesn't want to build in GitHub... Does it build locally?

gbiggs commented:
Not yet. This is still a work in progress.

MichaelOrlov commented on rosbag2_cpp/include/rosbag2_cpp/bag_events.hpp:
```diff
@@ -0,0 +1,134 @@
+
+  void execute_callbacks(const BagEvent event, BagEventCallbackBase::InfoPtr info)
+  {
+    std::for_each(
+      callbacks_.begin(),
+      callbacks_.end(),
+      [&info, &event](BagEventCallbackBase::SharedPtr & cb) {
+        if (cb->is_type(event)) {
+          cb->execute(info);
+        }
+      });
```
```suggestion
    for (auto & cb : callbacks_) {
      if (cb->is_type(event)) {
        cb->execute(info);
      }
    }
```
Simplification for more readable form.

MichaelOrlov commented on rosbag2_cpp/include/rosbag2_cpp/bag_events.hpp:
```diff
@@ -0,0 +1,134 @@
+#include <vector>
+
+#include "rclcpp/function_traits.hpp"
+
```
Extra empty line. Please remove it if there are no rational for this.

gbiggs replied:
Done in 961c7fc.

MichaelOrlov commented on rosbag2_transport/test/rosbag2_transport/mock_sequential_reader.hpp:
```diff
@@ -85,6 +85,11 @@ class MockSequentialReader : public rosbag2_cpp::reader_interfaces::BaseReaderIn
 
+  void add_event_callbacks(rosbag2_cpp::bag_events::ReaderEventCallbacks & /* callbacks */)
+  {
+    // TODO(gbiggs): Figure out how to mock this appropriately
```
I think we can do the similar as in regular sequential writer  
i.e. 
```
 if (callbacks.write_split_callback) {
    callback_manager_.add_event_callback(
      callbacks.write_split_callback,
      bag_events::BagEvent::WRITE_SPLIT);
  }
```
And call 
```
  file_number++;
  auto info = std::make_shared<bag_events::WriteSplitInfo>();
  info->closed_file = "ClosedFile" + std::to_string(file_number);
  info->opened_file = "OpenedFile" + std::to_string(file_number);
  callback_manager_.execute_callbacks(bag_events::BagEvent::WRITE_SPLIT, info);
```
in every `N` call of mocked write method.

gbiggs replied:
Done in 71c5349. Thanks for the tips.

MichaelOrlov commented on rosbag2_cpp/include/rosbag2_cpp/reader_interfaces/base_reader_interface.hpp:
```diff
@@ -57,6 +58,8 @@ class ROSBAG2_CPP_PUBLIC BaseReaderInterface
 
   virtual void seek(const rcutils_time_point_value_t & timestamp) = 0;
+
+  virtual void add_event_callbacks(bag_events::ReaderEventCallbacks & callbacks) = 0;
```
IMO. it will be much safer if `callbacks` parameter will be a const references.

gbiggs replied:
Done in db2fc1d.

MichaelOrlov commented on rosbag2_cpp/include/rosbag2_cpp/readers/sequential_reader.hpp:
```diff
@@ -101,6 +102,13 @@ class ROSBAG2_CPP_PUBLIC SequentialReader
+   *
+   * \param callbacks the structure containing the callback to add for each event.
+   */
+  void add_event_callbacks(bag_events::ReaderEventCallbacks & callbacks) override;
```
IMO. it will be much safer if `callbacks` parameter will be a const references.

gbiggs replied:
Done in db2fc1d.

MichaelOrlov reviewed (changes requested):
@gbiggs Thank you for implementing this feature.
I like idea with `EventCallbackManager`.

However it seems you forgot to add call for `execute_callbacks` to the `SequentialReader::load_next_file()`
IMO. It could be implemented as:
```
void SequentialReader::load_next_file()
{
  assert(current_file_iterator_ != file_paths_.end());
  auto info = std::make_shared<bag_events::BagSplitInfo>();
  info->closed_file = get_current_file();
  current_file_iterator_++;
  load_current_file();
  info->opened_file = get_current_file();
  callback_manager_.execute_callbacks(bag_events::BagEvent::READ_SPLIT, info);
}
```
I also have concern about potential performance degradation during bag split on recording. See my comments in code.

PS: Looking forward to see implementation for relevant unit tests.

gbiggs commented on rosbag2_cpp/include/rosbag2_cpp/reader.hpp:
```diff
@@ -179,6 +180,8 @@ class ROSBAG2_CPP_PUBLIC Reader final
     return *reader_impl_;
   }
 
+  void add_event_callbacks(bag_events::ReaderEventCallbacks & callbacks);
```
Done in a40ada3.

jhdcs reviewed (commented):
Just a quick comment. I'm also curious about a solution to the performance issue that @MichaelOrlov raised.

gbiggs commented on rosbag2_transport/test/rosbag2_transport/mock_sequential_reader.hpp:
```diff
@@ -56,6 +56,13 @@ class MockSequentialReader : public rosbag2_cpp::reader_interfaces::BaseReaderIn
   std::shared_ptr<rosbag2_storage::SerializedBagMessage> read_next() override
   {
     // filter_ was considered when incrementing num_read_ in has_next()
+    if (num_read_ % 5 == 0) {
```
It means the bag is "split" every 5 messages. I'll add a comment.

gbiggs commented:
I've added integration tests for the `Player` and `Recorder`. I'm still working on unit tests for `SequentialReader` and `SequentialWriter`.

gbiggs commented on rosbag2_transport/src/rosbag2_transport/recorder.cpp:
```diff
@@ -109,6 +110,19 @@ void Recorder::record()
+      auto message = rosbag2_interfaces::msg::WriteSplitEvent();
+      message.closed_file = info.closed_file;
+      message.opened_file = info.opened_file;
+      split_event_pub_->publish(message);
```
Done in 50ebfd3.

gbiggs commented:
I've pushed unit tests for the `SequentialWriter` and `SequentialReader`.

emersonknapp reviewed (approved):
LGTM - thanks for the revisions

emersonknapp commented:
Gist: https://gist.githubusercontent.com/emersonknapp/f8266fb628fa25dd88199616b529ea12/raw/20366e0739e0daff5b03c0d68c9a387a3566740b/ros2.repos
BUILD args: --packages-up-to rosbag2_cpp rosbag2_interfaces rosbag2_transport rosbag2_tests rosbag2
TEST args: --packages-select rosbag2_cpp rosbag2_interfaces rosbag2_transport rosbag2_tests rosbag2
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/9388
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15657)](http://ci.ros2.org/job/ci_linux/15657/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10335)](http://ci.ros2.org/job/ci_linux-aarch64/10335/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13323)](http://ci.ros2.org/job/ci_osx/13323/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15878)](http://ci.ros2.org/job/ci_windows/15878/)

jhdcs commented on rosbag2_cpp/include/rosbag2_cpp/bag_events.hpp:
```diff
@@ -0,0 +1,131 @@
+
+namespace rosbag2_cpp
+{
+
```
Would some comments in this file explaining what the events are and how to implement them be useful? I'm just thinking of future contributors to `rosbag2` that may not have been privy to the discussions on this. Comments may help to reduce onboarding time and lessen potential confusion. Maybe.

gbiggs replied:
Good catch! That file was completely devoid of comments. I've documented the entities in it in ad46aaa. Please take a look.

MichaelOrlov commented on rosbag2_transport/test/rosbag2_transport/mock_sequential_writer.hpp:
```diff
@@ -55,6 +55,15 @@ class MockSequentialWriter : public rosbag2_cpp::writer_interfaces::BaseWriterIn
     }
     messages_per_topic_[message->topic_name] += 1;
+    messages_per_file_ += 1;
+    if (messages_per_file_ == 5) {  // "Split" the bag every 5 messages
```
I would recommend to define some constant aka `constexpr size_t MAX_MESSAGES_PER_FILE = 5` somewhere in `mock_sequential_writer` and `mock_sequantial_reader` as public member and refer on it everywhere in tests.

gbiggs replied:
Done in 9e2cd6b.

MichaelOrlov replied:
I meant to define getters for maximum messages per file aka
```cpp
size_t max_messages_per_file();
```
and use it in `TEST_F(RosBag2PlayTestFixture, read_split_callback_is_called)` and `TEST_F(RecordIntegrationTestFixture, write_split_callback_is_called)` as well.

gbiggs replied:
I added getters for the constants, but the constant value is not used by the tests so there's nowhere to use them.

MichaelOrlov reviewed (changes requested):
@gbiggs Thank you for addressing my comments and proposals from previous round of review and new revisions.
Here is my second round of review.
I have proposal how to make `RosBag2PlayTestFixture::read_split_callback_is_called`  test non-flaky. See my proposals in code comments.

gbiggs commented on rosbag2_transport/src/rosbag2_transport/recorder.cpp:
```diff
@@ -83,6 +84,13 @@ Recorder::~Recorder()
+    event_publisher_thread_should_exit_ = true;
+  }
+  event_publisher_thread_wake_cv_.notify_all();
+  event_publisher_thread_->join();
```
How can the thread be constructed in the `Recorder` class constructor if it's on the stack? It runs a function that is a member of the `Recorder` class, but in the `Recorder` constructor, `this` is not yet valid.

MichaelOrlov replied:
Something like that 
```cpp
/ deferred start
std::thread thread;

// ...

// let's start now
thread = std::thread(functor, arg0, arg1);
```
Example taken from here https://stackoverflow.com/questions/7550362/in-c11-what-is-the-point-of-a-thread-which-does-not-represent-a-thread-of-ex

gbiggs replied:
Thanks for the tip! Done in b1fb4f8.

jhdcs commented on rosbag2_transport/src/rosbag2_transport/recorder.cpp:
```diff
@@ -141,6 +169,38 @@ void Recorder::record()
+      split_event_pub_->publish(message);
+    }
+
+    should_exit = event_publisher_thread_should_exit_;
```
Perhaps I'm missing something, but where is `event_publisher_thread_should_exit_` being set/modified after line 112? Is this going to be always `true`?

gbiggs replied:
Yes. The thread exists from the start of recording for the remaining lifetime of the `Recorder` and only exits when the recorder destructs. This is consistent with the way the `Recorder` class, once it starts recording, only stops recording when it destructs.

jhdcs reviewed (commented):
Looks great to me! Especially love the added documentation. I only had one area where I wasn't clear where something was being said, but I'm not confident that's not just me missing something.

emersonknapp commented:
I think we're good to move forward - have you addressed the yellow issues on OSX and Windows? Are you able to re-run CI yourself or would you like me to run it again?

gbiggs commented:
I haven't been able to address those issues. However a lot has changed since that CI run, so if you could run CI again that would be great.

MichaelOrlov reviewed (changes requested):
Suggestion for replacing magic number `size_t num_msgs_in_bag = 6;` with getter `max_messages_per_file()`  from `MockSequentialReader` and `MockSequentialWriter`

emersonknapp commented:
Gist: https://gist.githubusercontent.com/emersonknapp/ea90cff73b8a33a9cf4479ef2810e008/raw/20366e0739e0daff5b03c0d68c9a387a3566740b/ros2.repos
BUILD args: --packages-up-to rosbag2_cpp rosbag2_interfaces rosbag2_transport rosbag2_tests rosbag2
TEST args: --packages-select rosbag2_cpp rosbag2_interfaces rosbag2_transport rosbag2_tests rosbag2
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/9473
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15748)](http://ci.ros2.org/job/ci_linux/15748/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10429)](http://ci.ros2.org/job/ci_linux-aarch64/10429/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13412)](http://ci.ros2.org/job/ci_osx/13412/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15998)](http://ci.ros2.org/job/ci_windows/15998/)

gbiggs commented on rosbag2_transport/test/rosbag2_transport/test_play.cpp:
```diff
@@ -604,3 +605,64 @@ TEST_F(RosBag2PlayQosOverrideTestFixture, override_has_precedence_over_recorded)
+  }
+
+  auto prepared_mock_reader = std::make_unique<MockSequentialReader>();
+  prepared_mock_reader->prepare(messages, topic_types);
```
Done in db9f753.

MichaelOrlov reviewed (approved):
Looks good to me with green CI.

MichaelOrlov commented:
@gbiggs I see that `SequentialWriterTest.split_event_calls_callback` failing on windows. See details [here](https://ci.ros2.org/job/ci_windows/15998/testReport/)

emersonknapp commented on rosbag2_cpp/test/rosbag2_cpp/test_sequential_writer.cpp:
```diff
@@ -386,3 +386,62 @@ TEST_F(SequentialWriterTest, snapshot_mode_zero_cache_size_throws_exception)
+  }
+
+  ASSERT_TRUE(callback_called);
+  EXPECT_EQ(closed_file, storage_options_.uri + '/' + storage_options_.uri + "_0");
```
This is why the Windows test is failing - to create URIs always use a filesystem library rather than assuming path separators in your code. For this you can use `rcpputils::fs::path` object, which supports the `/` operator for concatenation. Something like

```suggestion
  std::string expected_closed = (rcpputils::fs::path(storage_options_.uri) / (storage_options_.uri + "_0")).string();
  EXPECT_EQ(closed_file, expected_closed);
```

gbiggs commented:
@emersonknapp Please re-run CI.

emersonknapp commented:
Gist: https://gist.githubusercontent.com/emersonknapp/f795bd70b6a8a5a0ee59c9ec1886f5f4/raw/20366e0739e0daff5b03c0d68c9a387a3566740b/ros2.repos
BUILD args: --packages-up-to rosbag2_cpp rosbag2_interfaces rosbag2_transport rosbag2_tests rosbag2
TEST args: --packages-select rosbag2_cpp rosbag2_interfaces rosbag2_transport rosbag2_tests rosbag2
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/9482
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15757)](http://ci.ros2.org/job/ci_linux/15757/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10438)](http://ci.ros2.org/job/ci_linux-aarch64/10438/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13422)](http://ci.ros2.org/job/ci_osx/13422/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=16009)](http://ci.ros2.org/job/ci_windows/16009/)

emersonknapp commented:
so close! just one clang warning

gbiggs commented:
@emersonknapp I don't have access to ci.ros2.org. I'll ask for it but it might take a little while to get at this time of year.

In the mean time, I've added the missing virtual destructor, so could you please run CI again?

emersonknapp commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15818)](http://ci.ros2.org/job/ci_linux/15818/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10496)](http://ci.ros2.org/job/ci_linux-aarch64/10496/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13482)](http://ci.ros2.org/job/ci_osx/13482/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=16103)](http://ci.ros2.org/job/ci_windows/16103/)

wep21 commented:
@gbiggs Could you fix conflicts and DCO?

gbiggs commented:
New CI:

- Linux [![Build Status](https://ci.ros2.org/job/ci_linux/16210/badge/icon)](https://ci.ros2.org/job/ci_linux/16210/)
- Linux-aarch64 [![Build Status](https://ci.ros2.org/job/ci_linux-aarch64/10837/badge/icon)](https://ci.ros2.org/job/ci_linux-aarch64/10837/)
- Windows [![Build Status](https://ci.ros2.org/job/ci_windows/16568/badge/icon)](https://ci.ros2.org/job/ci_windows/16568/)

gbiggs commented:
The failure in [the `Rpr__rosbag2__ubuntu_jammy_amd64` job](https://build.ros2.org/job/Rpr__rosbag2__ubuntu_jammy_amd64/5/console) appears to be a failure in `flake8` itself. That test passes when I run it locally. Perhaps it's a problem with the buildfarm?

gbiggs commented:
The failure in the GitHub action CI (in test_play_timing.cpp) doesn't happen in the manually-run CI, and I can't reproduce it locally either. Is it a flaky test?

clalancette commented:
> The failure in [the `Rpr__rosbag2__ubuntu_jammy_amd64` job](https://build.ros2.org/job/Rpr__rosbag2__ubuntu_jammy_amd64/5/console) appears to be a failure in `flake8` itself. That test passes when I run it locally. Perhaps it's a problem with the buildfarm?

I think this job just got unlucky and built during a time in Ubuntu 22.04 when `pycodestyle` was updated, but `flake8` wasn't.  Thus they were incompatible.  I'm going to kick the job again to see if it passes this time.

clalancette commented:
@ros-pull-request-builder retest this please.

gbiggs commented:
Trying to get this PR moving again, so here's some CI:

- Linux [![Build Status](https://ci.ros2.org/job/ci_linux/16790/badge/icon)](https://ci.ros2.org/job/ci_linux/16790/)
- Linux-aarch64 [![Build Status](https://ci.ros2.org/job/ci_linux-aarch64/11371/badge/icon)](https://ci.ros2.org/job/ci_linux-aarch64/11371/)
- Windows [![Build Status](https://ci.ros2.org/job/ci_windows/17200/badge/icon)](https://ci.ros2.org/job/ci_windows/17200/)

MichaelOrlov commented:
@gbiggs I would  be glad to see this PR moving forward and merged  when CI will be green.  
Could you please try to rebase and fix linter warnings one more time?

gbiggs commented:
Rebased and fixed the double include.

- Linux [![Build Status](https://ci.ros2.org/job/ci_linux/16864/badge/icon)](https://ci.ros2.org/job/ci_linux/16864/)
- Linux-aarch64 [![Build Status](https://ci.ros2.org/job/ci_linux-aarch64/11439/badge/icon)](https://ci.ros2.org/job/ci_linux-aarch64/11439/)
- Windows [![Build Status](https://ci.ros2.org/job/ci_windows/17269/badge/icon)](https://ci.ros2.org/job/ci_windows/17269/)

MichaelOrlov commented:
@gbiggs Some tests from play_for failed on CI Actions job https://build.ros2.org/job/Rpr__rosbag2__ubuntu_jammy_amd64/126/
Could you please take a look on failures or try to retest?

MichaelOrlov commented:
@gbiggs I was  not able to reproduce those failures in `test_play_for` from failing job https://build.ros2.org/job/Rpr__rosbag2__ubuntu_jammy_amd64/126/ locally even under the heavy load of my machine.
I used `stress -m 60 -c 12 -i 60`.
I have a suspicious that it was some collision between tests running in parallel on CI. 
We have topic name **"/topic1"** which is very common for other unit tests. I will consider to rename those topic name in future.

Merging this request then.

MichaelOrlov commented:
@gbiggs I want to defer release to the rolling until we will get merged at least #1032 and perhaps #1007

As regards to the possibility of backporting this PR to the `Humble`. 
I walked through changes in PR and it seems we are ok it terms of API/ABI compatibility. I didn't find that you are altering or removing any existent  API or member variable. Only adding new. In theory it should keep compatibility with previous version of library.

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-8-18-2022/27050/1
